fblock.go

```diff
@@ -8,24 +8,152 @@ import (
-	Transactions    []json.RawMessage `json:"transactions"`
+	BodyMR          string           `json:"bodymr"`          // Merkle root of the Factoid transactions which accompany this block.
+	PrevKeyMR       string           `json:"prevkeymr"`       // Key Merkle root of previous block.
+	PrevLedgerKeyMR string           `json:"prevledgerkeymr"` // Sha3 of the previous Factoid Block
+	ExchRate        int64            `json:"exchrate"`        // Factoshis per Entry Credit
+	DBHeight        int64            `json:"dbheight"`        // Directory Block height
+	Transactions    []*FBTransaction `json:"transactions"`    // The transactions inside the block
 
 	ChainID     string `json:"chainid,omitempty"`
 	KeyMR       string `json:"keymr,omitempty"`
 	LedgerKeyMR string `json:"ledgerkeymr,omitempty"`
 }
 
+// FBTransactions represents a single and valid transaction contained inside
+// of an FBlock.
+// The data has been rearranged from the raw json response to make it easier to work with.
+type FBTransaction struct {
+	TxID        string                     `json:"txid"` // hex string
+	Blockheight int64                      `json:"blockheight"`
+	Timestamp   time.Time                  `json:"timestamp"`
+	Inputs      []SignedTransactionAddress `json:"inputs"`
+	Outputs     []TransactionAddress       `json:"outputs"`
+	OutECs      []TransactionAddress       `json:"outecs"`
+}
+
+// TransactionAddress holds the relevant data for either an input or an output.
+// The amount is in either Factoshi (Input and Output) or EC (OutECs).
+// The RCDHash is the SHA256 hash of the RCD.
+// The address is the human readable address calculated from the RCDHash and type.
+type TransactionAddress struct {
+	Amount  uint64 `json:"amount"`
+	RCDHash string `json:"address"` // hex string
+	Address string `json:"useraddress"`
+}
+
+// SignedTransactionAddress contains a TransactionAddress along with the RCD and
+// cryptographic signatures specified by the RCD.
+type SignedTransactionAddress struct {
+	TransactionAddress
+	RCD        string   `json:"rcd"`        // hex string
+	Signatures []string `json:"signatures"` // slice of hex strings
+}
+
+// these two types are just used internally as intermediary holding structs
+// to transform the json response to the appropriate struct and back
+type rawFBTransaction struct {
+	TxID           string               `json:"txid"`
+	Blockheight    int64                `json:"blockheight"`
+	Millitimestamp int64                `json:"millitimestamp"`
+	Inputs         []TransactionAddress `json:"inputs"`
+	Outputs        []TransactionAddress `json:"outputs"`
+	OutECs         []TransactionAddress `json:"outecs"`
+	RCDs           []string             `json:"rcds"`
+	SigBlocks      []rawSigBlock        `json:"sigblocks"`
+}
+type rawSigBlock struct {
+	Signatures []string `json:"signatures"`
+}
+
+func (t *FBTransaction) MarshalJSON() ([]byte, error) {
```

Comment:
are MarshalJSON and UnmarshalJSON used anywhere?
And can you add unit test for those methods?

Reply:
They are used recursively by Marshalling/Unmarshalling FBlock.Transactions. UnmarshalJSON has unit test coverage due to TestUnmarshalFBlock but it appears MarshalJSON does not, which I'll add.